### PR TITLE
Only cache non-exact path mappings when opted-in by the service.

### LINKED ISSUE #967: Disable path mapping cache for prefix mapping
We currently cache a path mapping result as long as the response status is successful and the mapping does not have any path parameters: https://github.com/line/armeria/blob/8e6cbc36a92de7ae3e03a41531e7ab486cd0121a/core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java#L334-L338

This does not work well with prefix mapping which can map to infinit number of paths, because the mapping result will always have zero path parameters.

To address this issue, we could:

1. Disable caching for prefix mapping by default
2. Provide a way to enable path mapping caching for some special cases, e.g. `HttpFileService` would use prefix mapping but still could use cache because it will return successful response status for a limited number of paths.
   - Annotation? e.g. `@EnablePathMappingCache`
   - Interface?

     ```java
     public interface SomeInterfaceThatServiceWillImplement {
         boolean canCache(PathMapping, PathAndQuery);
     }
     ```

Thoughts?

/cc @hyangtack @huydx @anuraaga

### PR DESCRIPTION
**Pull request by anuraaga**
Currently, prefix mappings used with an infinite number of URLs (common for a frontend proxy) will cause cache-thrashing. While it can be disabled for these situations, defaulting to exact paths + our common services should be good for most cases either way.

Fixes #967

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/internal/PathAndQuery.java [hunk 1]
```diff
@@ -92,6 +93,11 @@ public void storeInCache(String rawPath) {
         }
     }
 
+    @VisibleForTesting
+    public static Cache<String, PathAndQuery> pathCache() {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/PathAndQuery.java]: Javadoc?

File: core/src/test/java/com/linecorp/armeria/server/HttpServerTest.java [hunk 2]
```diff
@@ -894,6 +913,33 @@ public void testTrailers() throws Exception {
         assertThat(res.trailingHeaders().get(AsciiString.of("foo")), is("bar"));
     }
 
+    @Test(timeout = 10000)
+    public void testExactPathCached() throws Exception {
+        org.assertj.core.api.Assertions.assertThat(client().get("/cached-exact-path")
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/HttpServerTest.java]: Would you mind just replacing hamcrest with assertj? 😉 I can send a follow-up PR if you do, though.

File: core/src/main/java/com/linecorp/armeria/internal/PathAndQuery.java (hunk 1 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/internal/PathAndQuery.java]: Ah, never mind. `PathAndQuery` is already an internal class.

File: core/src/main/java/com/linecorp/armeria/server/DecoratingService.java [hunk 3]
```diff
@@ -65,6 +66,11 @@ public void serviceAdded(ServiceConfig cfg) throws Exception {
         return result.isPresent() ? result : delegate.as(serviceType);
     }
 
+    @Override
+    public boolean shouldCachePath(PathAndQuery pathAndQuery, PathMapping pathMapping) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/DecoratingService.java]: `PathAndQuery` is an internal class. Could you replace it with two `String` parameters?

File: core/src/test/java/com/linecorp/armeria/server/HttpServerTest.java (hunk 2 above)
Comment (anuraaga) [reply core/src/test/java/com/linecorp/armeria/server/HttpServerTest.java]: Yep I was thinking of doing it myself in a follow up PR, will be much clearer

Comment (trustin) [reply core/src/test/java/com/linecorp/armeria/server/HttpServerTest.java]: Thanks!

File: core/src/main/java/com/linecorp/armeria/server/Service.java [hunk 4]
```diff
@@ -130,4 +132,12 @@ R decorate(Function<T, R> decorator) {
     default Service<I, O> decorate(DecoratingServiceFunction<I, O> function) {
         return new FunctionalDecoratingService<>(this, function);
     }
+
+    /**
+     * Returns whether the given {@code path} and {@code query }should be cached if the service's result is
```
Comment (imasahiro) [inline core/src/main/java/com/linecorp/armeria/server/Service.java]: nit. Move whitespace before `should`
`{@code query }should`
->
`{@code query} should`

Comment (anuraaga) [reply core/src/main/java/com/linecorp/armeria/server/Service.java]: Oops, thanks

**Comment by trustin** [comment]
Thanks, @minwoox

**Comment by trustin** [comment]
Thanks, @anuraaga !

**Pull request merged**